File: LC0669_Trim_BST/trim_bst.cpp
#include <cstddef>
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode* rTrimBST(TreeNode* root, int low, int high)
    {
        if (root == nullptr)
        {
            return NULL;
        }
        if (low <= root->val && root->val <= high)
        {
            root->left = this->rTrimBST(root->left, low, high);
            root->right = this->rTrimBST(root->right, low, high);
            return root;
        }
        // root is not within bounds
        // Case 1: root is a leaf
        if (root->left == nullptr && root->right == nullptr)
        {
            return NULL;
        }
        // Case 2: root has only one child
        if (root->left == nullptr)
        {
            return this->rTrimBST(root->right, low, high);
        }
        if (root->right == nullptr)
        {
            return this->rTrimBST(root->left, low, high);
        }
        // Case 3: root has both children
        if (root->val < low)
        {
            return this->rTrimBST(root->right, low, high);
        }
        else if (root->val > high)
        {
            return this->rTrimBST(root->left, low, high);
        }
        return NULL;
    }

    TreeNode* trimBST(TreeNode* root, int low, int high) {
        return this->rTrimBST(root, low, high);
    }
};

File: LC0669_Trim_BST/trim_bst.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def rTrimBST(self, root: TreeNode, low: int, high: int) -> TreeNode:
        if not root:
            return
        if low <= root.val <= high: # root is within boundaries
            root.left = self.rTrimBST(root.left, low, high)
            root.right = self.rTrimBST(root.right, low, high)
            return root
        else: # root is NOT within boundaries, and must be removed
            # Case 1: root is a leaf
            if (not root.left) and (not root.right):
                return None

            # Case 2: root has only one child
            if not root.left:
                return self.rTrimBST(root.right, low, high)
            if not root.right:
                self.rTrimBST(root.left, low, high)
                
            # Case 3: root has both children
            # Observation: If root.val < low, then its left subtree definitely must be removed (but we might keep its right subtree).
            # Conversely, if root.val > high, then its right subtree definitely must be removed (but we might keep its left subtree).
            if root.val < low:
                return self.rTrimBST(root.right, low, high)
            elif root.val > high:
                return self.rTrimBST(root.left, low, high)
            else:
                return None

    def trimBST(self, root: Optional[TreeNode], low: int, high: int) -> Optional[TreeNode]:
        return self.rTrimBST(root, low, high)